File: notes/fft/instfreq/compare.m
% 手動で瞬時周波数取得
% 
% instfreq 瞬時周波数の推定(Help)
% https://jp.mathworks.com/help/signal/ref/instfreq.html
% 「条件スペクトル モーメントとしての瞬時周波数と帯域幅」を一部改変

%% チャープ信号を生成
fs = 3e3;
t = 0:1/fs:2;
y = chirp(t,100,1,200,"quadratic");
y = vco(cos(2*pi*t),[0.1 0.4]*fs,fs);

% ---- 手動で瞬時周波数取得(ヒルベルト変換)
spec = fft(y);
spec(3000:6001) = 0;
spec = spec.* 2;
z2 = ifft(spec);
re = real(z2);
im = imag(z2);
iamp = sqrt(re.^2 + im.^2);
phase = atan2(im,re);
phase = unwrap(phase);
ifreq = (fs/(2*pi))*gradient(phase);
% ---- 手動で瞬時周波数取得(STFT経由)
[psPower,psFreq,psTime] = pspectrum(y,fs,'spectrogram');
tfd = psPower;
tfd = tfd/sum(tfd(:));
tfdSum = sum(tfd);
tmp = sum(psFreq.*tfd,1);
psMoment = tmp./tfdSum;

% ---- MALTABの場合
[z,tz] = instfreq(y,fs);
[a,ta] = tfsmoment(y,fs,1,Centralize=false);

%% ---- 以下比較
plot(tz,z,ta,a,'.',t,ifreq,'g',psTime,psMoment,'o');
legend("MATLAB-instfreq","MATLAB-tfsmoment","MY-hilbert","MY-stft");
% 一致していることがわかる

File: notes/fft/instfreq/compare.py
# %%
import numpy as np
from scipy.signal import chirp,stft,spectrogram
from scipy.fft import fft,ifft
import matplotlib.pyplot as plt
fs = 3000
t = np.linspace(0,2,fs*2)
y = chirp(t, f0=100, t1=1,f1=200,  method='linear') #note :  "quadratic"

# %%
spec = fft(y)
spec[3000:] = 0
spec = spec * 2
z = ifft(spec)
re = np.real(z)
im = np.imag(z)
phase = np.arctan2(im,re)
phase = np.unwrap(phase)
ifreq = (fs/(2*np.pi)) * np.gradient(phase)
plt.plot(t,ifreq)
# %%
freq,times,spec = stft(y,fs=fs,window="hann",nperseg=1024*2,noverlap=2017)
power = np.abs(spec) ** 2
tfd = power
tfd = tfd / np.sum(tfd)
tfdSum = np.sum(tfd,axis=0)
for i in range(tfd.shape[1]):
    tfd[:,i] *= freq
tmp = np.sum(tfd,axis=0)
moment = tmp / tfdSum
plt.pcolormesh(times, freq, power)
plt.show()
plt.plot(times,moment)
# %%
